File: question/neetcode/binary_tree/binary-tree-maximum-path-sum.py
from typing import Optional
import math

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        """
                1
            2       3
        4       5

        4, 5
        4+2+5
        4+2 5+2

        left+cur
        right+cur
        cur
        
        left+cur+right
        cur
        left
        right

        max of use cur +/without left/right
        max of anypath


        4,4
        5,5
        7,11
        3,3
        8,11
        """

        def dfs(node: Optional[TreeNode]):
            if not node:
                return -math.inf, -math.inf
            
            left_continue, left_max = dfs(node.left)
            right_continue, right_max = dfs(node.right)

            cur_continue = node.val + max(0, left_continue, right_continue)
            cur_max = max(cur_continue, left_max, right_max, node.val + left_continue + right_continue)

            return cur_continue, cur_max
        
        return dfs(root)[1]
    

File: question/neetcode/binary_tree/binary-tree-maximum-path-sum.cpp
#include<bits/stdc++.h>
using namespace std;

/**
* Definition for a binary tree node.
* struct TreeNode {
*     int val;
*     TreeNode *left;
*     TreeNode *right;
*     TreeNode() : val(0), left(nullptr), right(nullptr) {}
*     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
*     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
* };
*/
class Solution {
public:
    int dfs(TreeNode* node, int& a){
        if(not node) return -3000;

        int left = dfs(node->left,a);
        int right = dfs(node->right,a);

        int cur = node->val + max({0,left,right});
        a = max({a, cur, left, right, node->val+left+right});

        return cur;
    }

    int maxPathSum(TreeNode* root) {
        int a = -INT_MAX;

        dfs(root,a);
        return a;
    }
};